examples/boids_spatial3D/src/main.cu

```diff
@@ -195,62 +200,55 @@ FLAMEGPU_AGENT_FUNCTION(inputdata, flamegpu::MessageSpatial3D, flamegpu::Message
+        vec3Div(global_velocity_x, global_velocity_y, global_velocity_z, perceived_count);
+        //vec3Div(global_velocity_x, global_velocity_y, global_velocity_z, collision_count);
+
+        // Rule 1) Steer towards perceived centre of flock (Cohesion)
+        float steer_velocity_x = 0.f;
+        float steer_velocity_y = 0.f;
+        float steer_velocity_z = 0.f;
+        if (perceived_count > 0) {
+            const float STEER_SCALE = FLAMEGPU->environment.getProperty<float>("STEER_SCALE");
+            steer_velocity_x = (perceived_centre_x - agent_x) * STEER_SCALE;
+            steer_velocity_y = (perceived_centre_y - agent_y) * STEER_SCALE;
+            steer_velocity_z = (perceived_centre_z - agent_z) * STEER_SCALE;
+        }
+        velocity_change_x += steer_velocity_x;
+        velocity_change_y += steer_velocity_y;
+        velocity_change_z += steer_velocity_z;
+
+        // Rule 2) Match neighbours speeds (Alignment)
+        float match_velocity_x = 0.f;
+        float match_velocity_y = 0.f;
+        float match_velocity_z = 0.f;
+        if (perceived_count > 0) {
+            const float MATCH_SCALE = FLAMEGPU->environment.getProperty<float>("MATCH_SCALE");
+            match_velocity_x = global_velocity_x * MATCH_SCALE;
+            match_velocity_y = global_velocity_y * MATCH_SCALE;
+            match_velocity_z = global_velocity_z * MATCH_SCALE;
+        }
+        velocity_change_x += match_velocity_x - agent_fx;
+        velocity_change_y += match_velocity_y - agent_fy;
+        velocity_change_z += match_velocity_z - agent_fz;
+    } 
 
-    // Rule 1) Steer towards perceived centre of flock (Cohesion)
-    float steer_velocity_x = 0.f;
-    float steer_velocity_y = 0.f;
-    float steer_velocity_z = 0.f;
-    if (perceived_count > 0) {
-        const float STEER_SCALE = FLAMEGPU->environment.getProperty<float>("STEER_SCALE");
-        steer_velocity_x = (perceived_centre_x - agent_x) * STEER_SCALE;
-        steer_velocity_y = (perceived_centre_y - agent_y) * STEER_SCALE;
-        steer_velocity_z = (perceived_centre_z - agent_z) * STEER_SCALE;
-    }
-    velocity_change_x += steer_velocity_x;
-    velocity_change_y += steer_velocity_y;
-    velocity_change_z += steer_velocity_z;
-
-    // Rule 2) Match neighbours speeds (Alignment)
-    float match_velocity_x = 0.f;
-    float match_velocity_y = 0.f;
-    float match_velocity_z = 0.f;
-    if (collision_count > 0) {
-        const float MATCH_SCALE = FLAMEGPU->environment.getProperty<float>("MATCH_SCALE");
-        match_velocity_x = global_velocity_x * MATCH_SCALE;
-        match_velocity_y = global_velocity_y * MATCH_SCALE;
-        match_velocity_z = global_velocity_z * MATCH_SCALE;
-    }
-    velocity_change_x += match_velocity_x;
-    velocity_change_y += match_velocity_y;
-    velocity_change_z += match_velocity_z;
-
-    // Rule 3) Avoid close range neighbours (Separation)
```

Comment:
Rule 3 has just been entirely removed / replaced with just wall avoidance not avoidance of others?

Reply:
Rule through has moved into the message read loop as it is now dealing with individual agents rather than an aggregate value. I have added a comment to identify it.